docs/doc/60-contributing/03-rfcs/20221013-iceberg-external-table.md

```diff
@@ -0,0 +1,203 @@
+| INT
+| BIGINT
+| FLOAT
+| DOUBLE
+| DATE
+| TIMESTAMP
+| VARCHAR
+| ARRAY
+| OBJECT
+| VARIANT
+```
+
+- external-location (using S3 for example):
+
+```
+external-location ::=
+
+URL = 's3://<bucket-name>/<path-to-iceberg>'
+CONNECTION = (
+  ENDPOINT_URL = <endpoint_url>
+  ACCESS_KEY_ID = <access_key_id>
+  ACCESS_KEY_SECRET = <access_key_secret>
+  SESSION_TOKEN = <aws_session_token>
+)
+```
+
+### Accessing Iceberg Table's content
+
+```sql
+-- assuming users have created an external table named `iceberg_tbl`
+SELECT COUNT(*) FROM iceberg_tbl;
+```
+
+Joint query on normal table and Iceberg Table:
+
+```sql
+SELECT normal_tbl.book_name, iceberg_tbl.author FROM normal_tbl, iceberg_tbl WHERE normal_tbl.isbn = iceberg_tbl.isbn AND iceberg_tbl.sales > 100000;
+```
+
+On operating the table, all data remains still on the user-provided ends.
+
+### Time Travel
+
+Iceberg offers a list of snapshots and its timestamps. Time travelling on it is natrually.
+
+```sql
+SELECT ...
+FROM iceberg_table
+AT ( { SNAPSHOT => <snapshot_id> | TIMESTAMP => <timestamp> } );
+```
+
+#### Accessing Snapshot Metadata
+
+Users should be able to lookup the list of snapshot id in Iceberg:
+
+```sql
+SELECT snapshot_id FROM FUSE_SNAPSHOT('<database-name>', '<external-iceberg-table-name>');
+```
+
+```
+     snapshot_id
+---------------------
+ 1234567890123456789
+ 1234567890123456790
+(2 rows)
+```
+
+Current snapshot id the external table is reading will always be the newest. But users can using time travel with `AT`.
+
+## Reference-level explanation
+
+A new table engine `ICEBERG`, and options like `ENGINE_OPTION` for engine configuration and `external-location` in DDL will be added.
+
+### ICEBERG Engine
+
+ICEBERG is a table engine that enable users reading data from established Apache Iceberg endpoints. All table content and metadata of external table should remain in user-provided Iceberg data sources, in Iceberg's manner.
+
+The ENGINE will track the last commited snapshot, and should able to read from former snapshots.
+
+### ENGINE_OPTION
+
+Apache Iceberg has its own defined database and tables, but this external table only concerned about single tables in the Iceberg storage. Using ENGINE_OPTION users can specify which table in which database should the external access from Iceberg.
+
+### external-location
+
+No matter where the Iceberg is, like S3, GCS or OSS, if Databend support the storage, then the Iceberg should be accessable. Users need to tell databend where the Iceberg storage is and how should databend access the storage.
+
+## Drawbacks
+
+None
+
+## Rationale and alternatives
+
+### Create Table from Iceberg Snapshots
+
+Iceberg provides a snapshot capability and tables can be created from them.
+
+```sql
+CREATE TABLE [IF NOT EXISTS] [db.]table_name
+(
+    <column_name> <data_type> [ NOT NULL | NULL] [ { DEFAULT <expr> }],
+    <column_name> <data_type> [ NOT NULL | NULL] [ { DEFAULT <expr> }],
+    ...
+) ICEBERG_SNAPSHOT_LOCATION=
```

Comment:
Maybe we can specify it in `engine_options`?

Reply:
I think keeping tracking the latest committed Iceberg snapshot in the external tables is better. And users can access data in some snapshots, they can use `time travel`.

If users do want to clone data outside of `iceberg`, I prefer `COPY INTO`.

Reply:
I'm not talking about the snapshot-related logic about icebergs. I'm saying group all iceberg related configs into `engine_options`.

Reply:
This makes sense, I moved all remote location config into engine options, PTAL.